### ISSUE TITLE
Similarity and cross-encoder scores computed below float32 create spurious ties in low-precision inference

### ISSUE THREAD

**Issue by KisuYang** [UKPLab/sentence-transformers#3894]
### Summary

When a model runs in **float16/bfloat16** (e.g. `model_kwargs={"torch_dtype": torch.float16}`), sentence-transformers computes the final relevance score in that same low precision in two paths. Reduced-mantissa formats coarsely bucket values in the narrow score range, so **distinct query–document scores collapse into identical values ("spurious ties")**. Ranking then falls back to arbitrary tie-breaking (index order), which inflates the variance and bias of retrieval metrics and can even flip model rankings.

This is analyzed in [*Reliable Evaluation Protocol for Low-Precision Retrieval*](https://aclanthology.org/2026.acl-short.33) (Yang et al., ACL 2026). The remedy, **High-Precision Scoring (HPS), upcasts only the final scoring op to float32**, leaving the forward pass in low precision. It is value-preserving (every fp16/bf16 value is exactly representable in fp32), zero-parameter, and adds negligible cost.

MTEB recently merged the same two fixes: embeddings-benchmark/mteb#4933 (similarity primitives) and embeddings-benchmark/mteb#4934 (cross-encoder logits), from issue embeddings-benchmark/mteb#4932.

### Where it happens on current `main`

1. **Similarity primitives** — `sentence_transformers/util/tensor.py::_convert_to_tensor` already upcasts **sparse** tensors to float32, but leaves dense tensors at their incoming dtype. A fp16/bf16 model with `encode(..., convert_to_tensor=True)` returns low-precision embeddings, so `model.similarity()` / `util.cos_sim` / `dot_score` (and every evaluator built on them, e.g. `InformationRetrievalEvaluator`) scores in low precision. Note that `encode(..., convert_to_numpy=True)` upcasts **bfloat16** on conversion (numpy has no bf16) but keeps **float16** as fp16 numpy arrays, which are converted back to fp16 tensors when scoring.

2. **Cross-encoder scoring** — `CrossEncoder.predict` applies `activation_fn` (sigmoid) / softmax directly to the model's logits, i.e. in model dtype. The `convert_to_numpy` tail does call `.float()`, but only *after* the activation has already bucketized the (0, 1) probability range.

### Illustration

```python
import torch
q, d = torch.randn(1, 32), torch.randn(200, 32)
qn = torch.nn.functional.normalize
naive = qn(q.bfloat16(), dim=1) @ qn(d.bfloat16(), dim=1).T  # score in bf16
hps   = qn(q.bfloat16().float(), dim=1) @ qn(d.bfloat16().float(), dim=1).T  # HPS
print(torch.unique(naive).numel(), torch.unique(hps).numel())  # e.g. 184 vs 200

logits = torch.randn(200).bfloat16()  # cross-encoder logits
p_bf16 = torch.sigmoid(logits)
p_hps  = torch.sigmoid(logits.float())
print(torch.unique(p_bf16).numel(), torch.unique(p_hps).numel())  # e.g. ~150 vs 200
```

Fewer unique scores = spurious ties, broken arbitrarily during ranking.

### Proposed fix

Apply HPS at both sites, exactly as merged in MTEB:

- Upcast sub-float32 dense floating tensors to float32 in `_convert_to_tensor` — this mirrors the float32 upcast the same function already applies to sparse inputs.
- Upcast cross-encoder logits to float32 before the activation in `CrossEncoder.predict`.

Both are no-ops for float32 inputs, so full-precision results are unchanged. Under AMP/autocast training the affected ops are re-cast per autocast policy, so training behavior is effectively unchanged as well.

I've prepared two small, separate PRs and am opening them alongside this issue. Happy to adjust scope if maintainers prefer.

Disclosure: I'm the first author of the referenced paper; the same fixes were merged into MTEB this week (embeddings-benchmark/mteb#4933, embeddings-benchmark/mteb#4934).

**Comment by tomaarsen** [comment]
Hello!

I came to a very similar realisation in my upcoming Sentence Transformers v6.0 release prep (https://github.com/huggingface/sentence-transformers/pull/3794/changes/a26151975a622bead2c9874bffc7797402c01e1f), where I sum maxsim scores in fp32 as otherwise the bf16/fp16 results get "bucketed" together, resulting in a lot of ties. 

However, as I think this is a behaviour change that might surprise users (i.e. different output dtype), I will include this in the v6.0 release rather than the upcoming v5.7 release. Don't worry though, both will be rather soon. Thank you for these fixes, and nice work on the paper.

- Tom Aarsen

**Comment by KisuYang** [comment]
Hi @tomaarsen, thank you for the kind words and for taking the time to look into this!

Including it in v6.0 makes perfect sense. If it helps, I'd be happy to adjust the two PRs to fit your release plan.

Really glad to connect with you, and looking forward to the upcoming releases!
